velox/type/DecimalUtil.h

```diff
@@ -32,5 +32,38 @@ class DecimalUtil {
   /// Helper function to convert a decimal value to string.
   template <typename T>
   static std::string toString(const T& value, const TypePtr& type);
+
+  inline static std::optional<int128_t> rescaleWithRoundUp(
+      const int128_t& unscaledValue,
+      const uint8_t fromScale,
+      const uint8_t toPrecision,
+      const uint8_t toScale,
+      const bool nullOnFailure) {
+    auto rescaledValue = unscaledValue;
+    auto scaleDifference = toScale - fromScale;
+    if (scaleDifference >= 0) {
+      rescaledValue *= DecimalUtil::kPowersOfTen[scaleDifference];
+    } else {
+      scaleDifference = -scaleDifference;
+      rescaledValue /= DecimalUtil::kPowersOfTen[scaleDifference];
+      int128_t remainder =
+          unscaledValue % DecimalUtil::kPowersOfTen[scaleDifference];
+      if (unscaledValue >= 0 &&
+          remainder > DecimalUtil::kPowersOfTen[scaleDifference] / 2) {
+        rescaledValue++;
+      } else if (remainder < -DecimalUtil::kPowersOfTen[scaleDifference] / 2) {
+        rescaledValue--;
+      }
+    }
+    // Check overflowing
+    if (rescaledValue < -DecimalUtil::kPowersOfTen[toPrecision] ||
+        rescaledValue > DecimalUtil::kPowersOfTen[toPrecision]) {
+      if (nullOnFailure) {
+        return std::nullopt;
+      }
+      VELOX_FAIL();
```

Comment:
We can add the error message here.
Probably pass `fromPrecision` as well so that we can throw a proper message.